### PR TITLE
Avoid expensive directory scan for Illumina run completion

### PR DESCRIPTION
**Pull request by apmasell**
JIRA: GLT-2072

### CODE REVIEW

File: miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/IlluminaNotificationDto.java [hunk 1]
```diff
@@ -10,15 +10,16 @@
...
   private List<Integer> indexLengths;
   private int numCycles;
+  private int numReads;
   private Map<Integer, String> poolNames;
   private int readLength;
   private int scoreCycle;
-  private int bclCount;
 
   @Override
   public boolean equals(Object obj) {
```
Comment (sqwci) [inline miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/IlluminaNotificationDto.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 19 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)

**Comment by sqwci** [comment]
SonarQube analysis reported 7 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultIllumina.java#L137](https://github.com/TGAC/miso-lims/blob/f31f73f2a96f15699bc1176dc7246c9f36fe53ee/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L137): Refactor this method to reduce its Cognitive Complexity from 31 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L78](https://github.com/TGAC/miso-lims/blob/f31f73f2a96f15699bc1176dc7246c9f36fe53ee/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L78): Use the built-in formatting to construct this argument. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L138](https://github.com/TGAC/miso-lims/blob/f31f73f2a96f15699bc1176dc7246c9f36fe53ee/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L138): This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L150](https://github.com/TGAC/miso-lims/blob/f31f73f2a96f15699bc1176dc7246c9f36fe53ee/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L150): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)
1. ![MINOR][MINOR] [DefaultIllumina.java#L74](https://github.com/TGAC/miso-lims/blob/f31f73f2a96f15699bc1176dc7246c9f36fe53ee/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L74): Replace this "switch" statement by "if" statements to increase readability. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1301)
1. ![MINOR][MINOR] [DefaultIllumina.java#L175](https://github.com/TGAC/miso-lims/blob/f31f73f2a96f15699bc1176dc7246c9f36fe53ee/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L175): Refactor your code to get this URI from a customizable parameter. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1075)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sqwci** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultIllumina.java#L137](https://github.com/TGAC/miso-lims/blob/4ffd1b26760a1bf57650cc2c4fde5b67b2a2f5e3/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L137): Refactor this method to reduce its Cognitive Complexity from 31 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L138](https://github.com/TGAC/miso-lims/blob/4ffd1b26760a1bf57650cc2c4fde5b67b2a2f5e3/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L138): This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L150](https://github.com/TGAC/miso-lims/blob/4ffd1b26760a1bf57650cc2c4fde5b67b2a2f5e3/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L150): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)
1. ![MINOR][MINOR] [DefaultIllumina.java#L74](https://github.com/TGAC/miso-lims/blob/4ffd1b26760a1bf57650cc2c4fde5b67b2a2f5e3/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L74): Replace this "switch" statement by "if" statements to increase readability. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1301)
1. ![MINOR][MINOR] [DefaultIllumina.java#L175](https://github.com/TGAC/miso-lims/blob/4ffd1b26760a1bf57650cc2c4fde5b67b2a2f5e3/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L175): Refactor your code to get this URI from a customizable parameter. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1075)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java [hunk 2]
```diff
@@ -30,18 +30,27 @@
...
  * Scan an Illumina sequener's output using the Illumina Interop C++ library.
  *
  * This should work for all sequencer execept the Genome Analyzer and Genome Analyzer II.
  */
 public final class DefaultIllumina extends RunProcessor {
+  private static final LatencyHistogram directory_scan_time = new LatencyHistogram("runscanner_illumina_file_completness_time",
+      "The time to scan all the output files in a sequencer's directory to tell if it's finished.");
+
+  private static final Counter completness_method_success = Counter.build("runscanner_illumina_completness_check",
+      "The number of times a method was used to determine a run's completeness after sequencing").labelNames("method").register();
```
Comment (djcooke) [inline runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java]: :100:

File: runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java [hunk 3]
```diff
@@ -64,16 +75,19 @@ public static DefaultIllumina create(Builder builder, ObjectNode parameters) {
       case "CompletedAsPlanned":
         return Optional.of(HealthType.Completed);
       default:
-        log.debug("New Illumina completion status found: " + status);
+        log.debug("New Illumina completion status found: %s", status);
```
Comment (sqwci) [inline runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java]: MAJOR Use the built-in formatting to construct this argument. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2629)

**Comment by sqwci** [comment]
SonarQube analysis reported 7 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultIllumina.java#L137](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L137): Refactor this method to reduce its Cognitive Complexity from 31 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L138](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L138): This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L150](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L150): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)
1. ![MINOR][MINOR] [DefaultIllumina.java#L74](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L74): Replace this "switch" statement by "if" statements to increase readability. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1301)
1. ![MINOR][MINOR] [DefaultIllumina.java#L175](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L175): Refactor your code to get this URI from a customizable parameter. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1075)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java [hunk 4]
```diff
@@ -51,10 +60,12 @@
   private static final Pattern COMMA = Pattern.compile(",");
 
   private static final Predicate<String> BCL_FILENAME = Pattern.compile("^s_[0-9]*_[0-9]*\\.bcl(\\.gz)?").asPredicate();
+
   private static final Predicate<String> BCL_BGZF_FILENAME = Pattern.compile("^[0-9]*\\.bcl\\.bgzf").asPredicate();
 
   public static DefaultIllumina create(Builder builder, ObjectNode parameters) {
-    return new DefaultIllumina(builder);
+    return new DefaultIllumina(builder,
+        parameters.hasNonNull("checkOutput") ? parameters.get("checkOutput").asBoolean() : true);
```
Comment (djcooke) [inline runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java]: Please document this parameter in one of the readmes

**Comment by sqwci** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultIllumina.java#L137](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L137): Refactor this method to reduce its Cognitive Complexity from 31 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L138](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L138): This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MAJOR][MAJOR] [DefaultIllumina.java#L150](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L150): Either re-interrupt this method or rethrow the "InterruptedException". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)
1. ![MINOR][MINOR] [DefaultIllumina.java#L74](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L74): Replace this "switch" statement by "if" statements to increase readability. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1301)
1. ![MINOR][MINOR] [DefaultIllumina.java#L175](https://github.com/TGAC/miso-lims/blob/28037b74087c9382b50e5adfb661d04f9a5d9870/runscanner/src/main/java/uk/ac/bbsrc/tgac/miso/runscanner/processors/DefaultIllumina.java#L175): Refactor your code to get this URI from a customizable parameter. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1075)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**